tools/readaheadstat.py

```diff
@@ -0,0 +1,122 @@
+#
+# Based on readahead example from from BPF Performance Tools 
+# Book by Brendan Gregg
+#
+# Copyright (c) 2020 Suchakra Sharma <mail@suchakra.in>
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# 20-Aug-2020   Suchakra Sharma   Created this.
+
+from __future__ import print_function
+from bcc import BPF
+from time import sleep
+import ctypes as ct
+import argparse
+
+# arguments
+examples = """examples:
+    ./readaheadstat -d 20       # monitor for 10 seconds and generate stats 
+"""
+
+parser = argparse.ArgumentParser(
+    description="Monitor performance of read ahead cache",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-d", "--duration", type=int,
+    help="total duration to monitor for, in seconds")
+args = parser.parse_args()
+if not args.duration:
+    args.duration = 99999999
+
+# BPF program
+program = """
+#include <uapi/linux/ptrace.h>
+#include <linux/mm_types.h>
+
+BPF_HASH(flag, u32, u8);            // used to track if we are in do_page_cache_readahead()
+BPF_HASH(birth, struct page*, u64); // used to track timestamps of cache alloc'ed page
+BPF_ARRAY(pages);                   // increment/decrement readahead pages
+BPF_HISTOGRAM(dist);
+
+int entry__do_page_cache_readahead(struct pt_regs *ctx) {
+    u32 pid;
+    u8 one = 1;
+    pid = bpf_get_current_pid_tgid();
+    flag.update(&pid, &one);
+    return 0;
+}
+
+int exit__do_page_cache_readahead(struct pt_regs *ctx) {
+    u32 pid;
+    u8 zero = 0;
+    pid = bpf_get_current_pid_tgid();
+    flag.update(&pid, &zero);
+    return 0;
+}
+
+int exit__page_cache_alloc(struct pt_regs *ctx) {
+    u32 pid;
+    u64 ts;
+    struct page *retval = (struct page*) PT_REGS_RC(ctx);
+    u32 zero = 0; // static key for accessing pages[0]
+    pid = bpf_get_current_pid_tgid();
+    u8 *f = flag.lookup(&pid);
+    if (f != NULL && *f == 1) {
+        ts = bpf_ktime_get_ns();
+        birth.update(&retval, &ts);
+
+        u64 *count = pages.lookup(&zero);
+        if (count) (*count)++; // increment read ahead pages count
+    }
+    return 0;
+}
+
+int entry_mark_page_accessed(struct pt_regs *ctx) {
+    u64 ts, delta;
+    struct page *arg0 = (struct page *) PT_REGS_PARM1(ctx);
+    u32 zero = 0; // static key for accessing pages[0]
+    u64 *bts = birth.lookup(&arg0);
+    if (bts != NULL) {
+        delta = bpf_ktime_get_ns() - *bts;
+        dist.increment(bpf_log2l(delta/1000000));
+
+        u64 *count = pages.lookup(&zero);
+        if (count) (*count)--; // decrement read ahead pages count
+
+        birth.delete(&arg0); // remove the entry from hashmap
+    }
+    return 0;
+}
+"""
+
+b = BPF(text=program)
+b.attach_kprobe(event="__do_page_cache_readahead", fn_name="entry__do_page_cache_readahead")
+b.attach_kretprobe(event="__do_page_cache_readahead", fn_name="exit__do_page_cache_readahead")
+b.attach_kretprobe(event="__page_cache_alloc", fn_name="exit__page_cache_alloc")
```

Comment:
The same for all other kprobe based tracing. These functions may be inlined so the result from the tool may not represent what really happens in the system. I think this should be spelled out in the example and man page as warning, so if people double or cannot interpret their results, they should check vmlinux source and binary to confirm whether inlining happens or not. Analyzing dwarf probably can get similar information.

Reply:
Added notes in man page and example